### PR TITLE
Ros2 foxy refactor (rebased)

### PR DESCRIPTION
**Pull request by vilkun**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,10 +8,15 @@ https://rclUE.readthedocs.io/en/devel/
 
 - Ubuntu 20.04
 - Unreal Engine 4.27
+<<<<<<< HEAD
```
Comment (chalice-graeme) [inline README.md]: Seems like some merge artifacts in this file need to be cleaned up.

File: Scripts/ros.patch [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+<+>UTF-8
+===================================================================
+diff --git a/Source/ThirdParty/ros/include/rcutils/error_handling.h b/Source/ThirdParty/ros/include/rcutils/error_handling.h
+--- a/Source/ThirdParty/ros/include/rcutils/error_handling.h	(revision 584767ce5b22125a3cffe1e9e84618068f1431b4)
++++ b/Source/ThirdParty/ros/include/rcutils/error_handling.h	(date 1652721130821)
+@@ -88,7 +88,7 @@
+ } rcutils_error_state_t;
+ 
+ // make sure our math is right...
+-#if __STDC_VERSION__ >= 201112L
```
Comment (chalice-graeme) [inline Scripts/ros.patch:13]: What is the rationale behind this change? I understand that Clang doesn't define this for C++ builds, but does this static assert really need to run?

File: Scripts/update_ros.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+Notes:
+- 1 'rviz' error at the end, 1 'Failed to find ROS1', and several 'connext' errors during ros build, and 'rviz' during 'local_setup.bash' - is okay!
+- we rename SONAME of libs with version by 'patchelf' since UE4.27 can't deal with that in LinuxToolChain.cs, see more in RenameLibsWithVersion().
+'''
+
+import os, sys, shutil, re, time, subprocess
+from libs_utils import *
+
+# Main arguments:
+UE = '/home/vilkun/UE/UnrealEngine' # contains 'Engine' folder
```
Comment (chalice-graeme) [inline Scripts/update_ros.py]: Can we unify these paths and pass in the base path as a command-line parameter so that other people don't need to modify this script to build the ROS2 libraries?

File: Scripts/update_ros.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ros = '/home/vilkun/ros2_foxy' # '/opt/ros/foxy/lib'
+rclc = '/home/vilkun/work/build_foxy/rclc' # includes rclc_lifecycle
+ue_msgs = '/home/vilkun/work/build_foxy/UE_msgs' # check https://github.com/rapyuta-robotics/UE_msgs
+turtlebot3 = '/home/vilkun/work/turtlebot3-UE'
```
Comment (chalice-graeme) [inline Scripts/update_ros.py]: We build other internal projects with rclUE, so perhaps this can be made generic?

File: Scripts/update_ros.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    print('Grabbing libs...')
+    GrabLibs(rosBuild, rclUERosLib, allowed_spaces)
+    GrabLibs(rclcBuild, rclUERosLib, allowed_spaces)
+    GrabLibs(ue_msgsBuild, rclUERosLib, allowed_spaces)
+    CleanLibs(rclUERosLib, not_allowed_spaces)
+
+    RenameLibsWithVersion(rclUERosLib)
+    SetRPATH(rclUERosLib)
+    InvalidateBinaries(turtlebot3Binaries, rclUEBinaries, rclUEBuildCS)
```
Comment (chalice-graeme) [inline Scripts/update_ros.py]: See previous comment about us building other projects with rclUE, so perhaps this should be generic.

File: README.md [hunk 6]
```diff
@@ -24,6 +29,19 @@ This branch currently represents the release branch.
...
+To add or update part of ros2, for example 'rclc', branch 'foxy':
+1. Build ros2 https://docs.ros.org/en/foxy/Installation/Ubuntu-Development-Setup.html
+2. Clone part:
+git clone -b foxy https://github.com/ros2/rclc.git
+3. Make sure you installed colcon https://docs.ros.org/en/foxy/Tutorials/Colcon-Tutorial.html#create-a-workspace
+4. Build .so lib:
+. ~/ros2_foxy/install/local_setup.bash
+colcon build --packages-up-to --allow-overriding rclc
+(or without --allow-overriding, it's only in my case)
+5. Copy-replace includes and lib to appropriate folders to rclUE/Source/ThirdParty/ros2lib
```
Comment (chalice-graeme) [inline README.md]: Seems to be renamed rclUE/Source/ThirdParty/ros now?

File: Scripts/libs_utils.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
+import os, sys, shutil, re
+from subprocess import check_output
+
+# this script forces to link libraries again on next IDE run
+# (inside present a workaround for build.cs file, that's trick is to avoid deleting 'intermediate' folder at all)
+def InvalidateBinaries(turtlebot3Binaries, rclUEBinaries, rclUEBuildCS):
+    cwd = os.getcwd()
+
+    if os.path.isdir(turtlebot3Binaries):
```
Comment (chalice-graeme) [inline Scripts/libs_utils.py]: Seems like this could just be projectBinaries, since it is not specific to Turtlebot.

**Review by chalice-graeme — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (vilkun) [reply README.md]: ( will clean)

File: Scripts/ros.patch (hunk 2 above)
Comment (vilkun) [reply Scripts/ros.patch:13]: What do you mean? I created this patch with macro removing, because same patching was before my changes and was applyed to include folder. Also once in a while, not always, it helps to fix build at compile time

File: Scripts/update_ros.py (hunk 3 above)
Comment (vilkun) [reply Scripts/update_ros.py]: Yeah, okay, I will modify all home/vilkun as args

File: Scripts/update_ros.py (hunk 4 above)
Comment (vilkun) [reply Scripts/update_ros.py]: Same as above

File: Scripts/update_ros.py (hunk 5 above)
Comment (vilkun) [reply Scripts/update_ros.py]: Can't see path dependency here, think same as above

File: README.md (hunk 6 above)
Comment (vilkun) [reply README.md]: Some issue was during merging, I update this file

File: Scripts/libs_utils.py (hunk 7 above)
Comment (vilkun) [reply Scripts/libs_utils.py]: Yeah, I understand. Will change. all this paths I left, just for the latest stage of my refactor

File: Scripts/ros.patch (hunk 2 above)
Comment (chalice-graeme) [reply Scripts/ros.patch:13]: Ah, I did not realize this patch was already here.

**Review by chalice-graeme — APPROVED** [review]
